### PR TITLE
Introduce @QuarkusIntegrationTest

### LINKED ISSUE #13818: NativeImageTest with container-image extensions
**Description**

When using:
* `quarkus.native.container-build=true` (GraalVM/Mandrel native-image used from a docker image during NativeImageBuildStep)
* and `quarkus.container-image.build=true` (the application native executable is dockerized by JibProcessor)

it would be great if `@NativeImageTest` would be able to actually use that docker image to start the application, rather than trying to start the native executable inside/target from NativeImageBuildStep, because it may fail to start if incompatible with the host OS (i.e. building a Linux executable on a Windows laptop).

```xml
<properties>
  <quarkus.package.type>native</quarkus.package.type>
  <quarkus.native.container-build>true</quarkus.native.container-build>
  <quarkus.native.container-runtime>docker</quarkus.native.container-runtime>
  <quarkus.container-image.build>true</quarkus.container-image.build>
  <quarkus.container-image.push>false</quarkus.container-image.push>
</properties>
<dependencies>
  <dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-container-image-jib</artifactId>
  </dependency>
</dependencies>
```

This is also somehow related to #8654

**Implementation ideas**

I don't know if it is technically feasible for container-image-* extensions to change the behavior of NativeImageTest, which is core.
NativeImageLauncher might use TestContainers to start the containerized native image if `quarkus.container-image.build build=true`, rather than trying to launch the xxx-runner native executable ?
It would look for the image built during package phase, retrieving it from the docker daemon or the docker registry (depending on value of `quarkus.container-image.push`).

### PR DESCRIPTION
**Pull request by geoand**
The basic idea of this annotation is to test the final artifact produced by the Quarkus build, no matter what that is (a jar, a native binary or a container image).

TODO:

- [x] Polish code
- [x] Add tests
- [x] Add documentation

Resolves: #13818

### CODE REVIEW

**Comment by geoand** [comment]
This will conflict with https://github.com/quarkusio/quarkus/pull/14960, so I'll wait until the latter is merged

**Comment by geoand** [comment]
I have now added tests and cleaned things up a little.

@stuartwdouglas mind taking a look and letting me know what you think?

**Comment by ebramirez** [comment]
Hi @geoand, could this be extended to include testing `quarkus-picocli` applications?

**Comment by geoand** [comment]
I am not sure, I haven't really worked with the picocli extension.

What do you expect testing of a picocli application would look like?

**Comment by ebramirez** [comment]
I imagine that given a Picocli command like this:

```java
@Command
public class MyCommand {
    @Parameters( paramLabel = "PARAM", description = "App parameter", arity = "1" )
    Path param;

    @Option( names = { "--option", "-o" }, required = true )
    String option;

	/* Setters and Getters */
}
```

We could test it like this:

```java
@QuarkusIntegrationTest
class MyCommandTest {

    @Inject
    MyCommand myCommand;

    @Test
    void testValidOptions() {
        myCommand.setPath( "/my/path/param.txt" );
		myCommand.setOption( "my option" );

        final int exitStatus = myCommand.execute();

        assertThat( exitStatus ).isEqualTo( 0 );
    }
}
```

The `execute()` method from the injected command would pick up the annotated Options and Parameters and pass them to the appropriate launcher.

**Comment by geoand** [comment]
That can't work with this type of testing as the test launches a separate process and thus cannot interact with the application in any way other than standard ways to interact with a separate process - which is just a fancy way of say that `@Inject` can't work in these tests

**Comment by ebramirez** [comment]
Sorry for not explaining myself properly: the injected Command would not be the real command, just a proxy to collect the Options and Properties and pass them to the appropriate launcher (Jar, Native, Docker...).

Maybe something similar, but more automated, to using `@InjectMock` and `@RestClient`.

Does this make sense.

**Comment by geoand** [comment]
Thanks for the explanation.

That is probably possible and I'll consider it in the future.

File: test-framework/common/src/main/java/io/quarkus/test/common/ProcessReader.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Override
+    public void run() {
+        handleStart();
+        byte[] b = new byte[100];
+        int i;
+        try {
+            while ((i = inputStream.read(b)) > 0) {
+                String str = new String(b, 0, i, StandardCharsets.UTF_8);
+                System.out.print(str);
```
Comment (stuartwdouglas) [inline test-framework/common/src/main/java/io/quarkus/test/common/ProcessReader.java:23]: Its probably a massive corner case that we don't care about, but this won't display properly if you have a muti byte character that falls on the 100 byte mark.

File: test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusIntegrationTestExtension.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+            testResourceManager.init();
+            additionalProperties.putAll(testResourceManager.start());
+
+            String artifactType = quarkusArtifactProperties.getProperty("type");
+            if (artifactType == null) {
+                throw new IllegalStateException("Unable to determine the type of artifact created by the Quarkus build");
+            }
+            ArtifactLauncher launcher;
+            switch (artifactType) {
+                case "native": {
```
Comment (stuartwdouglas) [inline test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusIntegrationTestExtension.java:200]: This is ok for now, but I think long term we want the extension that contributes the artifact to also provide the launcher somehow to make this extensible.

Comment (geoand) [reply test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusIntegrationTestExtension.java:200]: Definitely. I was thinking of having a Service loader.
Once that need arises, I'll change the code ASAP

File: test-framework/common/src/main/java/io/quarkus/test/common/ProcessReader.java (hunk 1 above)
Comment (geoand) [reply test-framework/common/src/main/java/io/quarkus/test/common/ProcessReader.java:23]: Sounds like a corner case indeed 🙂.
But I'll keep it in mind and deal with it if it ever becomes a problem.

**Comment by geoand** [comment]
I'm gonna keep this PR in draft until we get DevDB support in as they will conflict and the DevDB support is more high value

**Comment by geoand** [comment]
At last the Windows tests on my fork as passing

**Comment by pipinet** [comment]
https://github.com/quarkusio/quarkus/issues/15198

**Comment by geoand** [comment]
Fixed conflict

**Comment by geoand** [comment]
GH seems to have messed up this PR.

I'll try to reopen later on or else just open a new one

**Comment by geoand** [comment]
Superseded by https://github.com/quarkusio/quarkus/pull/15227

**Pull request closed without merging**